### PR TITLE
Add pdf2 release management domain element processing

### PR DESCRIPTION
**Pull request by xephon2**
This is a pull request for processing elements of the new `Release Management` domain. To hook the processing, a fake element is needed.

``` xml
<frontmatter>
  <booklists>
    <toc/>
    <!-- This element does not exist but should exist to follow the given logic. -->
    <!--<changelist/>-->
    <!-- But it is possible to fake it. -->
    <booklist class="- map/topicref bookmap/changelist "/> 
    <tablelist/>
  </booklists>
</frontmatter>
```

A set of test files. The map in the test files contains the fake element, so no other hook or parameter is needed.

**Updated test files**:

[ReleaseManagement.zip](https://github.com/dita-ot/dita-ot/files/465913/ReleaseManagement.zip)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -121,4 +121,8 @@ $RECYCLE.BIN/
 # Windows shortcuts
 *.lnk
 
+# Log files
+*.log
```
Comment (jelovirt) [inline .gitignore]: Remove ignore changes

File: src/main/plugins/org.dita.odt/xsl/xslodt/common/topicmerge.xsl [hunk 2]
```diff
@@ -228,12 +228,13 @@ See the accompanying license.txt file for applicable licenses.
...
 			<xsl:when test="contains($class,' bookmap/booklist ')"/>
 			<xsl:when test="contains($class,' bookmap/booklists ')"/>
+            <xsl:when test="contains($class,' bookmap/changelist ')"/>
 			<xsl:when test="contains($class,' bookmap/colophon ')"/>
 			<xsl:when test="contains($class,' bookmap/dedication ')"/>
 			<xsl:when test="contains($class,' bookmap/figurelist ')"/>
 			<xsl:when test="contains($class,' bookmap/frontmatter ')"/>
 			<xsl:when test="contains($class,' bookmap/indexlist ')"/>
-			<xsl:when test="contains($class,' bookmap/tablelist ')"/>
+            <xsl:when test="contains($class,' bookmap/tablelist ')"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.odt/xsl/xslodt/common/topicmerge.xsl]: Just indentation change, revert

File: src/main/plugins/org.dita.odt/xsl/xslodt/common/topicmerge.xsl [hunk 3]
```diff
@@ -228,12 +228,13 @@ See the accompanying license.txt file for applicable licenses.
 			<xsl:when test="contains($class,' bookmap/bookabstract ')"/>
 			<xsl:when test="contains($class,' bookmap/booklist ')"/>
 			<xsl:when test="contains($class,' bookmap/booklists ')"/>
+            <xsl:when test="contains($class,' bookmap/changelist ')"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.odt/xsl/xslodt/common/topicmerge.xsl]: Correct indentation

File: src/main/plugins/org.dita.pdf2/build.xml [hunk 4]
```diff
@@ -0,0 +1,360 @@
+<?xml version="1.0" encoding="UTF-8"?><!--
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/build.xml]: build.xml is generated during integration, remove

File: src/main/plugins/org.dita.pdf2/build_template.xml [hunk 5]
```diff
@@ -244,9 +244,11 @@ with those set forth herein.
     <condition property="args.figurelink.style" value="NUMTITLE">
         <not><isset property="args.figurelink.style"/></not>
     </condition>
+
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/build_template.xml]: Revert indentation changes only

File: src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml [hunk 6]
```diff
@@ -328,16 +328,46 @@ See the accompanying license.txt file for applicable licenses.
...
+
      <!-- The heading string to put at the top of the List of Tables -->
      <variable id="List of Tables">الجداول</variable>
-     
+
      <!-- The heading string to put at the top of the List of Figures -->
      <variable id="List of Figures">الأشكال</variable>
 
+     <!-- The heading string to put at the top of the List of Changes -->
+     <variable id="List of Changes"></variable>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml]: @robander how do we want to handle untranslated strings in variable files. Having an English string in Arabic strings is not a particular good if the user doesn't speak Arabic.

File: src/main/plugins/org.dita.pdf2/xsl/fo/relmgmt-domain.xsl [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+                xmlns:fo="http://www.w3.org/1999/XSL/Format"
+                version="2.0">
+
+    <xsl:template match="*[contains(@class, ' relmgmt-d/change-item ')]">
+        <xsl:apply-templates/>
+    </xsl:template>
+
+    <xsl:template match="*[contains(@class, ' relmgmt-d/change-person ')]">
+        <xsl:text>change-person: </xsl:text>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/relmgmt-domain.xsl]: Don't hard code strings to stylesheets, use variable strings

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    </fo:block>
+  </xsl:template>
+
+  <xsl:template name="change-completed">
+    <xsl:param name="change-completed"/>
+    <fo:block>
+      <xsl:call-template name="getVariable"><xsl:with-param name="id" select="'Change Completed'"/></xsl:call-template>
+      <xsl:call-template name="getVariable"><xsl:with-param name="id" select="'ColonSymbol'"/></xsl:call-template>
+      <xsl:call-template name="getVariable"><xsl:with-param name="id" select="'figure-number-separator'"/></xsl:call-template>
+      <xsl:value-of select="format-date($change-completed,'[MNn] [D], [Y]')"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Date format string could be a variable strings, makes I18N easier

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                    <xsl:value-of select="$bookReleaseDateString"/>
+                    <xsl:value-of select="$colonSymbol"/>
+                    <xsl:value-of select="$separator"/>
+                    <xsl:value-of select="format-date($timeRangeEndAsDate,'[MNn] [D], [Y]')"/>
+                  </fo:block>
+                </fo:block>
+              </fo:table-cell>
+            </fo:table-row>
+            <!-- Iterate over all change-items -->
+            <xsl:for-each select="1 to (count($changeset))">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: This is a counter based for-loop and not suitable style in out XSLT. Use a normal for-each to process each `$changeset` node, thus `$currentChangeItem` becomes the current node.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    </fo:list-block>
+    <xsl:apply-templates select="*[contains(@class,' ditaot-d/ditaval-endprop ')]" mode="outofline"/>
+  </xsl:template>
+
+  <xsl:template match="*[contains(@class, ' relmgmt-d/change-item ')]/*" priority="10"/>
+
+  <xsl:template name="change-item">
+    <xsl:param name="change-item"/>
+    <xsl:for-each select="$change-item/*[contains (@class, ' relmgmt-d/change-organization ')][1]">
+      <xsl:call-template name="change-organization">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Instead of named templates, use templates with match. Either with modes or preferably just based on the element that is matched.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        <xsl:with-param name="id" select="'ColonSymbol'"/>
+      </xsl:call-template>
+    </xsl:variable>
+    <xsl:variable name="separator">
+      <xsl:call-template name="getVariable">
+        <xsl:with-param name="id" select="'figure-number-separator'"/>
+      </xsl:call-template>
+    </xsl:variable>
+
+    <xsl:for-each select="1 to (count($bookReleaseDates) - 1)">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Don't use counter based for-loops.

**Comment by jelovirt** [comment]
Squash commits at some point, no need to keep development history.

**Comment by jelovirt** [comment]
WRT `<booklist class="- map/topicref bookmap/changelist "/> `, we don't want to add fake elements like that. We need some other way to enable change history list generation.

File: src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml (hunk 6 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml]: @jelovirt and @robander : I added english and german strings and copied the english strings to all other language files.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 11 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: @jelovirt How should I iterate over a variable that is "an array" alternatively?

Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]:
```xml
<xsl:for-each select="$bookReleaseDates">
  ...
```

File: src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml (hunk 6 above)
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml]: @robander I think we should have a coding guideline for a case like this. What to do when adding a new variable string. Do you just add the English source text there and add e.g. a comment "FIXME: translate" or something to show that the content hasn't been translated yet?

Comment (robander) [inline src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml]: @jelovirt - that is what I did for the few items that I didn't have when adding languages last fall. Some of those, I couldn't actually see where the string was used (it was referenced in code but I couldn't find the markup to make it appear).

I think there is an argument either way (leave empty, or leave English). I went with English because that's more likely to be noticed and fixed if the string is ever needed. Empty strings will probably leave messed up output that wouldn't be noticed if doing a quick QA scan over the file.

**Comment by jelovirt** [comment]
Use `<booklist type="change-historylist"/>` to signal change history generation.

**Comment by robander** [comment]
:+1: I also like `<booklist type="change-historylist"/>`.

**Comment by xephon2** [comment]
@jelovirt done

**Comment by xephon2** [comment]
@jarno Do you think this PR could land with 2.3?

**Comment by jelovirt** [comment]
Started to merge this and fixing the issues it still had but had to stop. Some files had e.g. whitespace changes that needed to be reverted. However, the biggest problem was [`loc.xsl`][1]. It has so many XSLT anti-patterns that it cannot be merged as is. For example

```xslt
<xsl:for-each select="$changeset">
  <xsl:variable name="j" select="position()"/>
  <xsl:variable name="currentChangeItem" select="$changeset[$j]"/>
```

or

```xslt
<xsl:apply-templates mode="relmgmt-table" select="$currentChangeItem">
  <xsl:with-param name="change-item" select="$currentChangeItem"/>
</xsl:apply-templates>
```

The `loc.xsl` needs to be rewritten to follow XSLT coding approaches used by the rest of the project.

[1]: https://github.com/doctales/dita-ot/blob/relmgmt-pdf2/src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+                xmlns:opentopic-index="http://www.idiominc.com/opentopic/index"
+                xmlns:ot-placeholder="http://suite-sol.com/namespaces/ot-placeholder"
+                exclude-result-prefixes="opentopic opentopic-index dita2xslfo ot-placeholder xs"
+                version="2.0">
+
+  <!-- The variable changeset contains a list of all change-item elements. -->
+  <xsl:variable name="changeset" select="//*[contains (@class, ' relmgmt-d/change-item ')]"/>
+
+  <!-- The variable topicsWithChangeItem contains a list of all topics containing change-item elements. -->
+  <xsl:variable name="topicsWithChangeItem" select="//*[contains (@class, ' topic/topic ')][.//*[contains (@class, ' relmgmt-d/change-item ')]]"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Unused variable, remove

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+  <!-- Keep date format in a variable -->
+  <xsl:variable name="dateFormat">[MNn] [D], [Y]</xsl:variable>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Date format should be moved to common string variable as it's higly language/locale dependent.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+          <xsl:with-param name="id" select="'List of Changes'"/>
+        </xsl:call-template>
+      </fo:marker>
+      <xsl:call-template name="getVariable">
+        <xsl:with-param name="id" select="'List of Changes'"/>
+      </xsl:call-template>
+    </fo:block>
+  </xsl:template>
+  
+  <xsl:template match="//*[contains(@class, ' bookmap/bookmeta ')]/*[contains(@class, ' topic/critdates ')]">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: No need to start pattern with `//`, remove

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+      </xsl:call-template>
+    </xsl:variable>
+    <xsl:variable name="separator">
+      <xsl:call-template name="getVariable">
+        <xsl:with-param name="id" select="'figure-number-separator'"/>
+      </xsl:call-template>
+    </xsl:variable>
+
+    <!--<xsl:for-each select="1 to (count($bookReleaseDates) - 1)">-->
+    <xsl:for-each select="$bookReleaseDates ">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: `$bookReleaseDates` is two attribute nodes. Replace index based for-loop with current node based approach.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+                    <xsl:value-of select="$bookReleaseDateString"/>
+                    <xsl:value-of select="$colonSymbol"/>
+                    <xsl:value-of select="$separator"/>
+                    <xsl:value-of select="format-date($timeRangeEndAsDate, $dateFormat)"/>
+                  </fo:block>
+                </fo:block>
+              </fo:table-cell>
+            </fo:table-row>
+            <!-- Iterate over all change-items -->
+            <xsl:for-each select="$changeset">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Replace index based for-loop with `xsl:for-each` and current node approach.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+              <xsl:if test="$timeRangeStart != ''
+                            and $currentChangeItemChangeCompleted != ''
+                            and $timeRangeEnd">
+                <xsl:variable name="currentChangeItemChangeCompletedAsDate" select="xs:date($currentChangeItemChangeCompleted)"/>
+                <xsl:if test="($timeRangeStartAsDate &lt;= $currentChangeItemChangeCompletedAsDate) and ($currentChangeItemChangeCompletedAsDate &lt;= $timeRangeEndAsDate)">
+                  <fo:table-row xsl:use-attribute-sets="releaseManagementTable.changeItem.row">
+                    <fo:table-cell xsl:use-attribute-sets="releaseManagementTable.changeItem.cell">
+                      <fo:block xsl:use-attribute-sets="releaseManagementTable.changeItem.content">
+                        <xsl:apply-templates mode="relmgmt-table" select="$currentChangeItem">
+                          <xsl:with-param name="change-item" select="$currentChangeItem"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: No need to pass `$currentChangeItem` as parameter since it will be the current node.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+              </xsl:if>
+            </xsl:for-each>
+          </fo:table-body>
+        </fo:table>
+      </fo:block>
+    </xsl:for-each>
+  </xsl:template>
+
+  <xsl:template match="*[contains(@class, ' relmgmt-d/change-item ')]" mode="relmgmt-table">
+    <xsl:param name="change-item"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: `change-item` parameter is the same node as the current node, refactor to use current node

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+    <xsl:for-each select="$change-item/*[contains (@class, ' relmgmt-d/change-organization ')][1]">
+      <xsl:apply-templates mode="relmgmt-table-change-organization"
+                           select="$change-item/*[contains (@class, ' relmgmt-d/change-organization ')][1]"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: No need to loop through single `change-organization` element and then apply templates for it, just replace with

```xslt
<xsl:apply-templates mode="relmgmt-table-change-organization"
+                           select="*[contains (@class, ' relmgmt-d/change-organization ')][1]"/>
```

File: src/main/plugins/org.dita.pdf2/cfg/fo/attrs/loc-attr.xsl [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+                xmlns:fo="http://www.w3.org/1999/XSL/Format"
+                xmlns:rx="http://www.renderx.com/XSL/Extensions"
+                version="2.0">
+
+    <xsl:attribute-set name="releaseManagementTable">
+    </xsl:attribute-set>
+
+    <xsl:attribute-set name="releaseManagementTable.bookRelease.row">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/cfg/fo/attrs/loc-attr.xsl]: Tables/rows/cells should use attribute sets from normal table styling, e.g. `use-attribute-sets="tbody.row"` in this case.

**Comment by xephon2** [comment]
Thanks a lot for your comments, @jelovirt. I'm going to refactor the code as soon as possible.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 12 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 13 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: I copy the variable into the var files and follow the patterns from https://en.wikipedia.org/wiki/Date_format_by_country, OK?

Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 14 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 17 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 16 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 15 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 18 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 19 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/cfg/fo/attrs/loc-attr.xsl (hunk 20 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/cfg/fo/attrs/loc-attr.xsl]: done
However the output looks quite boring now. Should I apply any other default attribute-sets?
loc

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 21]
```diff
@@ -74,52 +66,35 @@
...
-                            and $timeRangeEnd">
-                <xsl:variable name="currentChangeItemChangeCompletedAsDate" select="xs:date($currentChangeItemChangeCompleted)"/>
-                <xsl:if test="($timeRangeStartAsDate &lt;= $currentChangeItemChangeCompletedAsDate) and ($currentChangeItemChangeCompletedAsDate &lt;= $timeRangeEndAsDate)">
-                  <fo:table-row xsl:use-attribute-sets="releaseManagementTable.changeItem.row">
-                    <fo:table-cell xsl:use-attribute-sets="releaseManagementTable.changeItem.cell">
-                      <fo:block xsl:use-attribute-sets="releaseManagementTable.changeItem.content">
-                        <xsl:apply-templates mode="relmgmt-table" select="$currentChangeItem">
-                          <xsl:with-param name="change-item" select="$currentChangeItem"/>
+            <xsl:for-each select="$change-items">
+              <xsl:if test="./*[contains (@class, ' relmgmt-d/change-completed ')] != ''">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: `./` part is redundant here, remove

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 22]
```diff
@@ -74,52 +66,35 @@
...
+                <xsl:variable name="change-completed" select="xs:date(./*[contains (@class, ' relmgmt-d/change-completed ')])"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Remove `./`

File: src/main/plugins/org.dita.odt/xsl/xslodt/common/topicmerge.xsl (hunk 3 above)
Comment (jelovirt) [inline src/main/plugins/org.dita.odt/xsl/xslodt/common/topicmerge.xsl]: This feature is only for PDF, not ODT. Remove this change

File: src/main/java/org/dita/dost/util/Constants.java [hunk 23]
```diff
@@ -85,6 +85,7 @@
     public static final DitaClass BOOKMAP_BOOKRIGHTS = new DitaClass("- topic/data bookmap/bookrights ");
     public static final DitaClass BOOKMAP_BOOKTITLE = new DitaClass("- topic/title bookmap/booktitle ");
     public static final DitaClass BOOKMAP_BOOKTITLEALT = new DitaClass("- topic/ph bookmap/booktitlealt ");
+    public static final DitaClass BOOKMAP_CHANGELIST = new DitaClass("- map/topicref bookmap/changelist ");
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/util/Constants.java]: Not needed, remove

File: src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml [hunk 24]
```diff
@@ -1,20 +1,20 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml]: Remove whitespace-only changes

File: src/main/plugins/org.dita.pdf2/xsl/fo/topic2fo.xsl [hunk 25]
```diff
@@ -92,13 +92,18 @@ See the accompanying license.txt file for applicable licenses.
     <xsl:import href="markup-domain.xsl"/>
     <xsl:import href="../../cfg/fo/attrs/xml-domain-attr.xsl"/>
     <xsl:import href="xml-domain.xsl"/>
+    <xsl:import href="../../cfg/fo/attrs/relmgmt-domain-attr.xsl"/>
+    <xsl:import href="relmgmt-domain.xsl"/>
+
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/topic2fo.xsl]: Remove whitespace

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+                <fo:block xsl:use-attribute-sets="tbody.row.entry__content">
+                  <xsl:value-of select="$bookReleaseDateString"/>
+                  <xsl:value-of select="$colonSymbol"/>
+                  <xsl:value-of select="$separator"/>
+                  <xsl:value-of select="format-date(., $dateFormat)"/>
+                </fo:block>
+              </fo:table-cell>
+            </fo:table-row>
+            <xsl:for-each select="$change-items">
+              <xsl:if test="./*[contains (@class, ' relmgmt-d/change-completed ')] != ''">
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Remove `./`

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 27]
```diff
@@ -0,0 +1,178 @@
...
+                <xsl:variable name="change-completed" select="xs:date(./*[contains (@class, ' relmgmt-d/change-completed ')])"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Remove `./`

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 28]
```diff
@@ -0,0 +1,178 @@
...
+                <xsl:if test="($previousDate &lt;= $change-completed)
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: The logic is hard to follow. Filtering should be done earlier if it's needed. The idea that previous date is used here is something I just don't understand. Could this not be handled with sorting at a higher level?

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 29]
```diff
@@ -0,0 +1,178 @@
...
+                               and ($change-completed &lt;= $date)">
+                  <fo:table-row xsl:use-attribute-sets="tbody.row">
+                    <fo:table-cell xsl:use-attribute-sets="tbody.row.entry">
+                      <fo:block xsl:use-attribute-sets="tbody.row.entry__content">
+                        <xsl:apply-templates mode="relmgmt-table" select=".">
+                          <xsl:with-param name="change-item" select="."/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Not used, remove

File: src/main/java/org/dita/dost/util/Constants.java (hunk 23 above)
Comment (xephon2) [inline src/main/java/org/dita/dost/util/Constants.java]: done

File: src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml (hunk 24 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/cfg/common/vars/ar.xml]: I reverted all var files and then re-added the new vars.

File: src/main/plugins/org.dita.pdf2/xsl/fo/topic2fo.xsl (hunk 25 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/topic2fo.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 26 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 27 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 28 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: I now only collect `change-item` elements that have a `change-completed` child node.

```xslt
<!-- The variable change-items contains a list of all change-item elements that have a change-completed date. -->
  <xsl:variable name="change-items" select="//*[contains (@class, ' relmgmt-d/change-item ')][*[contains (@class, ' relmgmt-d/change-completed ')]]"/>
```

Further on, I removed the now obsolete if clause:

```xslt
<xsl:if test="./*[contains (@class, ' relmgmt-d/change-completed ')] != ''">
```

Unfortunately I don't know how to remove the previous date and I don't understand what's wrong with it. If you want to collect changes that occured in a specific time range, you need a start date and an end date. And the start date of a time range here is always the previous release of a book.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl (hunk 29 above)
Comment (xephon2) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: done

**Comment by jelovirt** [comment]
This PR has merge conflicts, the code isn't finished enough yet, and lacks a champion from the reviewer team. Moving target to 3.x.

**Comment by xephon2** [comment]
I think, I'm done.  :tada:

**Comment by xephon2** [comment]
🍂 **push**

**Comment by xephon2** [comment]
@robander How about adding a milestone to this?

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+  </xsl:template>
+  
+  <xsl:template match="*[contains(@class, ' bookmap/bookmeta ')]/*[contains(@class, ' topic/critdates ')]">
+
+    <!-- Create a list of all book releases -->
+    <xsl:variable name="bookReleaseDates" select="*[contains (@class, ' topic/created ')]/@date |
+                                                  *[contains (@class, ' topic/revised ')]/@modified"/>
+    <xsl:variable name="bookReleaseDateString">
+      <xsl:call-template name="getVariable">
+        <xsl:with-param name="id" select="'Book Release Date'"/>
```
Comment (robander) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: It would simplify the code to combine these into one `getVariable` call instead of using variable for the date + variable for colon + variable for separator + value of date.

In the variable file, you can refer to other variables, so the "Book Release Date" variable itself can be a combination of (date text) + colon + separator.

You can also pass the date in as a parameter when getting the "Book Release Date" variable -- so that you end up only getting + rendering a single OT variable, and never need to set `xsl:variable`:
```
<xsl:call-template name="getVariable">
  <xsl:with-param name="id" select="'Book Release Date'">
  <xsl:with-param name="params">
    <date><xsl:value-of select="format-date(., $dateFormat)"/></date>
  </xsl:with-param>
</xsl:call-template>
```

This will work with an updated string value:
```
<variable id="Book Release">Book Release<variable id="ColonSymbol"/><variable id="figure-number-separator"/><param ref-name="date"/></variable>
```

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+    </xsl:for-each>
+  </xsl:template>
+
+  <xsl:template match="*[contains(@class, ' relmgmt-d/change-item ')]" mode="relmgmt-table">
+    <xsl:apply-templates mode="relmgmt-table"/>
+  </xsl:template>
+
+  <xsl:template match="*[contains (@class, ' relmgmt-d/change-organization ')]" mode="relmgmt-table">
+    <fo:block>
+      <xsl:call-template name="getVariable"><xsl:with-param name="id" select="'Change Organization'"/></xsl:call-template>
```
Comment (robander) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: Suggest simplifying this as above, and with other combined variables.

File: src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+                  <xsl:value-of select="$bookReleaseDateString"/>
+                  <xsl:value-of select="$colonSymbol"/>
+                  <xsl:value-of select="$separator"/>
+                  <xsl:value-of select="format-date(., $dateFormat)"/>
+                </fo:block>
+              </fo:table-cell>
+            </fo:table-row>
+            <xsl:for-each select="$change-items">
+              <xsl:variable name="change-completed" select="xs:date(*[contains (@class, ' relmgmt-d/change-completed ')])"/>
+              <xsl:if test="($previousDate &lt;= $change-completed)
```
Comment (robander) [inline src/main/plugins/org.dita.pdf2/xsl/fo/loc.xsl]: This confused me when I started creating more test data.

I have a set of change data that has a date before the original release date (change made in 2016, first release from the book meta is in 2017 -- and nothing showed up for the change. I am guessing this section is responsible.

**Comment by robander** [comment]
Just as an FYI -- I've added these elements into my general NLS Test files, currently assuming that the change list would be generated with `<booklist outputclass="changelist">`, and assuming that it is still tied to `<critdates>` metadata. I'm not sure if this is useful or not right now, but longer term this enables testing of the output for every supported language. https://github.com/robander/metadita.nlstest

**Comment by jelovirt** [comment]
Development on this PR has stalled and I'm closing it as inactive. I suggest this is separated to a separate DITA-OT plugin.

**Comment by xephon2** [comment]
This feature is now provided by the [org.doctales.release-notes](https://github.com/doctales/org.doctales.release-notes) plugin.
